### PR TITLE
Issue #15027: Add support for LITERAL_WHEN in whitespaceAroundCheck

### LINKED ISSUE #15027: Add Check Support for Java 21 Pattern Matching for Switch Syntax: WhitespaceAround
child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation: https://checkstyle.org/checks/whitespace/whitespacearound.html#WhitespaceAround
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="WhitespaceAround"/>
    </module>
</module>
➜  src cat Test.java 
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {

    void test(Object obj) {
        int y = 0;
        if(y >= 2) { }      // violation
        switch (obj) {
            case ColoredPoint(int x, _, _) when(x >= 2) -> { }   // expected violation
            default -> { }
        }
    }

    void test2(Object obj) {
        int y=0; // 2 violations, no space before and after '='
        switch (obj) {
            case ColoredPoint(int x, _, _)when(x >= 2) -> { } // (expected) 2 violations, no space before and after 'when'
            case ColoredPoint(int x, _, _) when(x >= 0)  -> { }  // (expected) 1 violation, no space after 'when'
            default -> { }
        }
    }
}
➜  src javac --enable-preview --release 21 Test.java
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:8:9: 'if' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:16:14: '=' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:16:14: '=' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 3 errors.

```

---

Describe what you want in detail

violation on `when`.

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #15027 

Diff Regression config: https://gist.githubusercontent.com/mahfouz72/afd57aa6e51ca161b7927acd959e5211/raw/2c9cfc2b9a7daad757c39ec06d41a75da0c95730/check.xml
Diff Regression patch config: https://gist.githubusercontent.com/mahfouz72/33967bf86e4a29cafa429771b0e95808/raw/f607f60b0a9a5ad9ecf1e13fea9e31d4dddc32da/check_patch.xml

### CODE REVIEW

**Comment by mahfouz72** [comment]
github, generate report

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: We don't need this here, let's remove this comment.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java:18]: Same as https://github.com/checkstyle/checkstyle/pull/15304#discussion_r1683362869, let's find other tokens to put here.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java
Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: done

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java:18]: done

**Review by nrmancuso — APPROVED** [review]

**Comment by nrmancuso** [comment]
We are also missing checks in this PR

**Review by nrmancuso — APPROVED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: Show me that when is covered here.

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: can you restate. I don't understand. what do you mean by "covered here"

Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: @mahfouz72 what would happen if we don't exclude this token? It should not appear in our code until we migrate to java 21.

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: there is a test that validate that all checks tokens are used in google config https://github.com/checkstyle/checkstyle/blob/42158b1fbad0ee3ef2d60a246635c805910c4057/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java#L391

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:183]: So we should add it to the config or exclude it like I did to make the test passes

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: added `LITERAL_WHEN` to the config.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: @mahfouz72 What is your understanding of what it means to put the token here at this location?

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: Ideally, we should have all checks tokens in our checkstyle config and we have a test to enforce that. This list is to exclude some tokens from being added in the config (suppress the test for this specific token). an example for this `GENERIC_START` which should be in the config of the whitespaceAround check but we didn't add it there because we have a separate check for it so we added it in this exclude list

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: My understanding at first is why we would need to add LITERAL_WHEN in the config? we don't use this syntax in our code what will we gain if we add it in the config? that's why I excluded it

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: Please correct me if anything is wrong and clarify this comment https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1685632454

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: So this area is to document what tokens are excluded from our config as well as the other configs. When we add these tokens here, we document why. The comment that is next to the token you added is `these are covered by GenericWhitespaceCheck`. To me, that comment coveys why we are adding this token here. So my question at the time was, how does `GenericWhitespaceCheck` cover this token for another check to satisfy why we aren't adding it.

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: ahh, The comment was misleading I didn't want to say that it applies to `LITERAL_WHEN`. actually there was another comment (https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1683363896)    `// 'when' keyword can be used by newer java versions only` but I removed it based on @nrmancuso review he was probably was trying to say to remove the token at all from the list but i removed the comment only

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: Okay now we are good I added the token to the config and removed it from the exclude list. I just was confused why we added to our config a token that we are not using. but as @nrmancuso said [here](https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1685742936) there should be no issues with this. we will use it when we migrate to java 17

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java:392]: @nrmancuso Is there a reason we don't make this default? We should our reasoning to the first post in the issue.

It seems like this won't hurt anything unless its used and if it is used, it would only be on a project which supports it. I am not seeing any negatives yet on why this can't be defaulted.

File: src/main/resources/google_checks.xml
Comment (rnveach) [inline src/main/resources/google_checks.xml:130]: @nrmancuso @romani ping.

@mahfouz72 If we are changing google checks, we need documentation that the google style guide supports this change.
Basically what specific rule supports this.
https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html

For our config, we like to use as much as possible, but only if we really agree with the rule being applied to our repo. If we don't agree with it, we usually still add it but set it's violations to be ignored so we are still atleast no exception testing it.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/resources/google_checks.xml
Comment (rnveach) [reply src/main/resources/google_checks.xml:130]: @Zopsss ping as well

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java:22]: Indentation. Switch is wrong location.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java:10]: Add some examples of a new line before/after `when`.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java:392]: Good point, let’s make this default

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java:392]: @mahfouz72 Please make this default also. This should also revert `checkstyle-checks.xml` to what it was before.

File: src/main/resources/google_checks.xml
Comment (Zopsss) [reply src/main/resources/google_checks.xml:130]: [4.6.2 Horizontal whitespace](https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace) section defines rules for Whitespace around code. The documentation does not specifically mentions anything about the `when` keyword but the google-java-format adds whitespace around `when`. Here's the formatted code:

```java
record ColoredPoint(int p, int x, String c) {}

record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) {}

public class Test {

  void test(Object obj) {
    int y = 0;
    if (y >= 2) {} // violation
    switch (obj) {
      case ColoredPoint(int x, _, _) when (x >= 2) -> {} // expected violation
      default -> {}
    }
  }

  void test2(Object obj) {
    int y = 0; // 2 violations, no space before and after '='
    switch (obj) {
      case ColoredPoint(int x, _, _) when (x
          >= 2) -> {} // (expected) 2 violations, no space before and after 'when'
      case ColoredPoint(int x, _, _) when (x
          >= 0) -> {} // (expected) 1 violation, no space after 'when'
      default -> {}
    }
  }
}
```

`when` was introduced in Java 21, the last major update for style guide was before Java 21 released so it does not contains rule for newly introduced keywords. 

But I think we should add support for this keyword, it is ensures basic spacing between `when` and other code and it is overall improving the coding style

Comment (Zopsss) [reply src/main/resources/google_checks.xml:130]: If we agree with this keyword then I think it's we should test cases for it in [InputWhitespaceAroundBasic.java](https://github.com/checkstyle/checkstyle/blob/master/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAroundBasic.java) file.

Comment (rnveach) [reply src/main/resources/google_checks.xml:130]: @Zopsss Even though it is not specifically mentioned, are there any parts of the style which indicate this spacing follows the heart of the guide? Unfortunately we can't go by what "looks good" since that is subjective.

This is the closest I am seeing
> Separating any reserved word

`when` is a keyword like the others
https://docs.oracle.com/javase/specs/jls/se22/html/jls-3.html#jls-3.9
But it isn't require to have a `(` after it, but the google documentation also clearly states the spaces should only appear in very specific places.

@romani @nrmancuso What do you think?

Comment (Zopsss) [reply src/main/resources/google_checks.xml:130]: > But it isn't require to have a `(` after it

I see. If that's the case then based on style guide, we don't need any kind rule for `when` as it does not mentions anything explicitly about it. 

Though the formatter does adds spacing around it and I also think we should add support for this token to maintain consistent spacing. I'll be waiting for other maintainers opinion on this.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java:395]: done added to the default tokens and revered checkstyle-checks.xml

File: src/main/resources/google_checks.xml
Comment (mahfouz72) [reply src/main/resources/google_checks.xml:130]: > Separating any reserved word, such as if, for or catch, from an open parenthesis (() that follows it on that line

`when` is a reserved word and it may have an open parenthesis so I think google style should support this token

> But it isn't require to have a ( after it

why does this information affect our decision? Yes it is not required but It still may have an open parenthesis

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java]: done

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundLiteralWhen.java:17]: done

File: src/main/resources/google_checks.xml
Comment (rnveach) [reply src/main/resources/google_checks.xml:130]: @mahfouz72 Please read https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace .
Google config follows style guide and style guide seems to be very strict in this case on what is supposed to happen with spacing.

Comment (romani) [reply src/main/resources/google_checks.xml:130]: @mahfouz72 , please use formatter https://github.com/google/google-java-format on input to confirm that `when` in spaced on both sides.

Comment (mahfouz72) [reply src/main/resources/google_checks.xml:130]: @Zopsss used the formatter and confirmed that it adds spaces around `when` https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1686824725

Comment (Zopsss) [reply src/main/resources/google_checks.xml:130]: > @mahfouz72 , please use formatter https://github.com/google/google-java-format on input to confirm that `when` in spaced on both sides.

Formatter adds spaces on both sides of `when` but there's no spacing rule defined for Java 21 keywords in the style guide because style guide is more older than the Java 21 release.

Comment (romani) [reply src/main/resources/google_checks.xml:130]: @mahfouz72 , please add example of "when" to new Input at folder
https://github.com/checkstyle/checkstyle/tree/master/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting
and create test method at https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/HorizontalWhitespaceTest.java

Comment (mahfouz72) [reply src/main/resources/google_checks.xml:130]: https://github.com/checkstyle/checkstyle/pull/15359

Comment (rnveach) [reply src/main/resources/google_checks.xml:130]: Same state as other PR. @nrmancuso @romani

Comment (romani) [reply src/main/resources/google_checks.xml:130]: Yes, this PR is above my expectations.
Thanks a lot.

**Comment by mahfouz72** [comment]
@rnveach @nrmancuso we should unblock this aslo

**Comment by nrmancuso** [comment]
> @rnveach @nrmancuso we should unblock this aslo

done, let's party

**Review by rnveach — APPROVED** [review]

**Pull request merged**
